src/python/zquantum/core/distance_measures.py

```diff
@@ -0,0 +1,108 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import numpy as np
+
+
+def compute_rbf_kernel(x_i, y_j, sigma):
+    """ Compute the gaussian (RBF) kernel matrix K, with K_ij = exp(-gamma |x_i - y_j|^2) and gamma = 1/(2*sigma).
+
+        Args:
+            x_i (np.array): Samples A (integers).
+            y_j (np.array): Samples B (integers).
+            sigma (float): The bandwidth of the gaussian kernel.
+
+        Returns:
+            np.ndarray: The gaussian kernel matrix.
+    """
+    exponent = np.abs(x_i[:, None] - y_j[None, :]) ** 2
+    try:
+        gamma = 1.0 / (2 * sigma)
+    except ZeroDivisionError as error:
+        print("Handling run-time error:", error)
+        raise
+    K = np.exp(-gamma * exponent)
+    return K
+
+
+def compute_multi_rbf_kernel(x_i, y_j, sigmas):
+    """ Compute the multi-gaussian (RBF) kernel matrix K, with K_ij = 1/c Sum_n[exp(-gamma_n |x_i - y_j|^2)] and gamma = 1/(2*sigma).
+
+        Args:
+            x_i (np.array): Samples A (integers).
+            y_j (np.array): Samples B (integers).
+            sigmas (np.array): The list of bandwidths of the multi-gaussian kernel.
+
+        Returns:
+            np.ndarray: The gaussian kernel matrix.
+    """
+    exponent = np.abs(x_i[:, None] - y_j[None, :]) ** 2
+    K = 0.0
+    for sigma in sigmas:
+        try:
+            gamma = 1.0 / (2 * sigma)
+        except ZeroDivisionError as error:
+            print("Handling run-time error:", error)
+            raise
+        K = K + np.exp(-gamma * exponent)
+    return K / len(sigmas)
+
+
+def compute_mmd(
+    target_distribution, measured_distribution, distance_measure_parameters,
+):
+    """ Compute the squared Maximum Mean Discrepancy (MMD) distance measure between between a target bitstring distribution
+    and a measured bitstring distribution.
+    Reference: arXiv.1804.04168.
+
+        Args:
+            target_distribution (BitstringDistribution): The target bitstring probability distribution.
+            measured_distribution (BitstringDistribution): The measured bitstring probability distribution.
+            distance_measure_parameters (dict): dictionary containing the relevant parameters for the maximum mean discrepancy, i.e.:
+                                                - sigma: the bandwidth parameter used to compute the gaussian kernel.
```

Comment:
I would describe `distance_measure_parameters` as a separate section from `Args`, to make it easier to read (especially in cases with more parameters).
It also lacks information about default value.
sth like:
```
distance_measure_parameters:
    sigma (float): the bandwidth parameter used to compute the gaussian kernel. The default value is 1.0 .
```

Reply:
Changed it [here](https://github.com/zapatacomputing/z-quantum-core/pull/63/commits/6f169a7f4ff15f0441b6de762463849ed13bf2ad)

Reply:
Updated also for the clipped log likelihood cost function [here](https://github.com/zapatacomputing/z-quantum-core/pull/63/commits/a8f0187dd7936ac3611261cbb14b0adfeba0a1a9)